### PR TITLE
Add handling for NoPia object creation in control flow graph.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Comment by AlekseyTs** [comment]
>             if (operation.Initializer != null)

In [https://github.com/dotnet/roslyn/pull/27460](https://github.com/dotnet/roslyn/pull/27460) this code is already extracted into a helper, once that is merged I will also change VisitNoPiaObjectCreation to take advantage of that helper instead of duplicating the logic.

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:4723 in abdc6ab. [](commit_id = abdc6ab2a5086196822e702b14c8c8bc0b368ff1, deletion_comment = False)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -4941,12 +4961,14 @@ IOperation popTarget(IOperation originalTarget, ImmutableArray<IOperation> argum
 
         public override IOperation VisitObjectOrCollectionInitializer(IObjectOrCollectionInitializerOperation operation, int? captureIdForResult)
         {
-            throw ExceptionUtilities.Unreachable;
+            Debug.Assert(false, "This code path should not be reachable.");
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Debug.Fail

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -4941,12 +4961,14 @@ IOperation popTarget(IOperation originalTarget, ImmutableArray<IOperation> argum
...
+            return MakeInvalidOperation(operation.Syntax, operation.Type, ImmutableArray<IOperation>.Empty);
         }
 
         public override IOperation VisitMemberInitializer(IMemberInitializerOperation operation, int? captureIdForResult)
         {
-            throw ExceptionUtilities.Unreachable;
+            Debug.Assert(false, "This code path should not be reachable.");
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Debug.Fail

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INoPiaObjectCreationOperation.cs [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+    Statements (4)
+        IFlowCaptureOperation: 0 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'x')
+          Value: 
+            IParameterReferenceOperation: x (OperationKind.ParameterReference, Type: ITest33) (Syntax: 'x')
+
+        IFlowCaptureOperation: 1 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'new ITest33  { y }')
+          Value: 
+            INoPiaObjectCreationOperation (OperationKind.None, Type: ITest33) (Syntax: 'new ITest33  { y }')
+              Initializer: 
+                null
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INoPiaObjectCreationOperation.cs]: Why is the initializer null here? I would expect it to have the collection initializer value.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_INoPiaObjectCreationOperation.cs]: > Why is the initializer null here? I would expect it to have the collection initializer value. 

Object creation is broken into individual operations: an actual instance creation followed by individual initializations.

---
In reply to: [193145223](https://github.com/dotnet/roslyn/pull/27476#discussion_r193145223) [](ancestors = 193145223)

**Review by 333fred — APPROVED** [review]
LGTM (iteration 1)

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@mavasani, @dotnet/analyzer-ioperation Please review

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 4]
```diff
@@ -4692,6 +4692,26 @@ public override IOperation VisitInvocation(IInvocationOperation operation, int?
             return (visitedInstance, visitedArguments);
         }
 
+        internal override IOperation VisitNoPiaObjectCreation(INoPiaObjectCreationOperation operation, int? argument)
+        {
+            // Initializer is removed from the tree and turned into a series of statements that assign to the created instance
+            IOperation initializedInstance = new NoPiaObjectCreationOperation(initializer: null, semanticModel: null, operation.Syntax, operation.Type,
+                                                                              operation.ConstantValue, IsImplicit(operation));
+
+            if (operation.Initializer != null)
```
Comment (mavasani) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: I believe @333fred is [pulling out](https://github.com/dotnet/roslyn/pull/27460/files#diff-2c0a650e29227920ce474a68cd354d0eL4704) similar blocks of code from individual Visit methods into `HandleObjectOrCollectionInitializer`. You may want to wait for his change to go in and simplify this code?

**Review by mavasani — APPROVED** [review]

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Yes, Aleksey said he'd move correct that once my PR is merged, which it is now. https://github.com/dotnet/roslyn/pull/27476#issuecomment-394776622

**Pull request merged**